src.related/pbmplus/pbm/pbmreduce.1
.TH pbmreduce 1 "02 August 1989"
.SH NAME
pbmreduce - read a portable bitmap and reduce it N times
.SH SYNOPSIS
.B pbmreduce
.RB [ -floyd | -fs | -threshold ]
.RB [ -value
.IR val ]
.I N
.RI [ pbmfile ]
.SH DESCRIPTION
Reads a portable bitmap as input.
Reduces it by a factor of
.IR N ,
and produces a portable bitmap as output.
.PP
.I pbmreduce
duplicates a lot of the functionality of
.I pgmtopbm;
you could do something like
.B pnmscale | pgmtopbm,
but
.I pbmreduce
is a lot faster.
.PP
.I pbmreduce
can be used to "re-halftone" an image.
Let's say you have a scanner that only produces black&white, not
grayscale, and it does a terrible job of halftoning (most b&w scanners
fit this description).
One way to fix the halftoning is to scan at the highest possible
resolution, say 300 dpi, and then reduce by a factor of three or
so using
.IR pbmreduce .
You can even correct the brightness of an image, by using the
.B -value
flag.
.SH OPTIONS
.PP
By default, the halftoning after the reduction is done via
boustrophedonic Floyd-Steinberg error diffusion; however, the
.B -threshold
flag can be used to specify simple thresholding.  This gives better
results when reducing line drawings.
.PP
The
.B -value
flag alters the thresholding value for all quantizations.
It should be a real number between 0 and 1.
Above 0.5 means darker images; below 0.5 means lighter.
.PP
All flags can be abbreviated to their shortest unique prefix.
.SH "SEE ALSO"
pnmenlarge(1), pnmscale(1), pgmtopbm(1), pbm(5)
.SH AUTHOR
Copyright (C) 1988 by Jef Poskanzer.
.\" Permission to use, copy, modify, and distribute this software and its
.\" documentation for any purpose and without fee is hereby granted, provided
.\" that the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.  This software is provided "as is" without express or
.\" implied warranty.


CHANGES
- GISLIB 
  - G_warning(), G_fatal_error() - send mail only if eviroment variable
                                   GRASS_ERROR_MAIL is set
  - G_debug() - new, print debugging info on selected level
  - G_plot_icon() - new, plot selected mark
- DBMI 
  - dbmscap file is not used, drivers are searched in dist.$ARCH/driver/db
  - new dbf driver
- vector library completely rewritten (multilayer, multiformat, 3D)
- new make system


src.related/pbmplus/pbm/pbmreduce.c
/* pbmreduce.c - read a portable bitmap and reduce it N times
**
** Copyright (C) 1989 by Jef Poskanzer.
**
** Permission to use, copy, modify, and distribute this software and its
** documentation for any purpose and without fee is hereby granted, provided
** that the above copyright notice appear in all copies and that both that
** copyright notice and this permission notice appear in supporting
** documentation.  This software is provided "as is" without express or
** implied warranty.
*/

#include "pbm.h"

void main( argc, argv )
int argc;
char *argv[];
    {
    FILE *ifp;
    register bit **bitslice, *newbitrow, *nbP;
    int argn, n, rows, cols, format, newrows, newcols;
    int row, col, limitcol, subrow, subcol, count, direction;
    char *usage = "[-floyd|-fs | -threshold] [-value <val>] N [pbmfile]";
    int halftone;
#define QT_FS 1
#define QT_THRESH 2
#define SCALE 1024
#define HALFSCALE 512
    long threshval, sum, *thiserr, *nexterr, *temperr;

    pbm_init( &argc, argv );

    argn = 1;
    halftone = QT_FS;
    threshval = HALFSCALE;

    while ( argn < argc && argv[argn][0] == '-' && argv[argn][1] != '\0' )
	{
	if ( pm_keymatch( argv[argn], "-fs", 2 ) ||
	     pm_keymatch( argv[argn], "-floyd", 2 ) )
	    halftone = QT_FS;
	else if ( pm_keymatch( argv[argn], "-threshold", 2 ) )
	    halftone = QT_THRESH;
	else if ( pm_keymatch( argv[argn], "-value", 2 ) )
	    {
	    float f;

	    argn++;
	    if ( argn == argc || sscanf( argv[argn], "%f", &f ) != 1 ||
		 f < 0.0 || f > 1.0 )
		pm_usage( usage );
	    threshval = f * SCALE;
	    }
	else
	    pm_usage( usage );
	argn++;
	}

    if ( argn == argc )
	pm_usage( usage );
    if ( sscanf( argv[argn], "%d", &n ) != 1 )
	pm_usage( usage );
    if ( n < 2 )
	pm_error( "N must be greater than 1", 0,0,0,0,0 );
    argn++;

    if ( argn == argc )
	ifp = stdin;
    else
	{
	ifp = pm_openr( argv[argn] );
	argn++;
	}

    if ( argn != argc )
	pm_usage( usage );

    pbm_readpbminit( ifp, &cols, &rows, &format );
    bitslice = pbm_allocarray( cols, n );

    newrows = rows / n;
    newcols = cols / n;
    pbm_writepbminit( stdout, newcols, newrows, 0 );
    newbitrow = pbm_allocrow( newcols );

    if ( halftone == QT_FS )
	{
	/* Initialize Floyd-Steinberg. */
	thiserr = (long *) malloc( ( newcols + 2 ) * sizeof(long) );
	nexterr = (long *) malloc( ( newcols + 2 ) * sizeof(long) );
	if ( thiserr == 0 || nexterr == 0 )
	    pm_error( "out of memory", 0,0,0,0,0 );

	srandom( (int) time( 0 ) );
	for ( col = 0; col < newcols + 2; col++ )
	    thiserr[col] = ( random( ) % SCALE - HALFSCALE ) / 4;
	    /* (random errors in [-SCALE/8 .. SCALE/8]) */
	}
    direction = 1;

    for ( row = 0; row < newrows; row++ )
	{
	for ( subrow = 0; subrow < n; subrow++ )
	    pbm_readpbmrow( ifp, bitslice[subrow], cols, format );

	if ( halftone == QT_FS )
	    for ( col = 0; col < newcols + 2; col++ )
		nexterr[col] = 0;
	if ( direction )
	    {
	    col = 0;
	    limitcol = newcols;
	    nbP = newbitrow;
	    }
	else
	    {
	    col = newcols - 1;
	    limitcol = -1;
	    nbP = &(newbitrow[col]);
	    }

	do
	    {
	    sum = 0;
	    count = 0;
	    for ( subrow = 0; subrow < n; subrow++ )
		for ( subcol = 0; subcol < n; subcol++ )
		    if ( row * n + subrow < rows && col * n + subcol < cols )
			{
			count += 1;
			if ( bitslice[subrow][col * n + subcol] == PBM_WHITE )
			    sum += 1;
			}
	    sum = ( sum * SCALE ) / count;

	    if ( halftone == QT_FS )
		sum += thiserr[col + 1];

	    if ( sum >= threshval )
		{
		*nbP = PBM_WHITE;
		if ( halftone == QT_FS )
		    sum = sum - threshval - HALFSCALE;
		}
	    else
		*nbP = PBM_BLACK;

	    if ( halftone == QT_FS )
		{
		if ( direction )
		    {
		    thiserr[col + 2] += ( sum * 7 ) / 16;
		    nexterr[col    ] += ( sum * 3 ) / 16;
		    nexterr[col + 1] += ( sum * 5 ) / 16;
		    nexterr[col + 2] += ( sum     ) / 16;
		    }
		else
		    {
		    thiserr[col    ] += ( sum * 7 ) / 16;
		    nexterr[col + 2] += ( sum * 3 ) / 16;
		    nexterr[col + 1] += ( sum * 5 ) / 16;
		    nexterr[col    ] += ( sum     ) / 16;
		    }
		}
	    if ( direction )
		{
		col++;
		nbP++;
		}
	    else
		{
		col--;
		nbP--;
		}
	    }
	while ( col != limitcol );

	pbm_writepbmrow( stdout, newbitrow, newcols, 0 );

	if ( halftone == QT_FS )
	    {
	    temperr = thiserr;
	    thiserr = nexterr;
	    nexterr = temperr;
	    direction = ! direction;
	    }
	}

    pm_close( ifp );

    exit( 0 );
    }
